graql/reasoner/rule-interaction.feature

```diff
@@ -54,62 +54,111 @@ Feature: Rule Interaction Resolution
-    $charlie isa person, has name "charlie";
-    $dennis isa person, has name "dennis";
-
-    (attendee: $bob, speaker: $alice) isa conference;
-    (member: $alice, member: $bob, host: $alice) isa party;
-
-    (attendee: $charlie, speaker: $dennis) isa conference;
-    (member: $charlie, member: $dennis, host: $charlie) isa party;
-    """
-  When materialised database is completed
+  Scenario: when rules are similar but different the reasoner knows to distinguish the rules
+    Given for each session, graql define
+      """
+      define
+
+      person
+      plays lesson:teacher,
+      plays lesson:student,
+      owns tag;
+
+      lesson sub relation,
+      relates teacher,
+      relates student;
+
+      tag sub attribute, value string;
+
+      rule tag-teacher-leaders:
+      when {
+        $x isa person;
+        $y isa person;
+        (student: $x, teacher: $y) isa lesson;
+        (member: $x, member: $y, leader: $y) isa teams;
+      } then {
+        $y has tag "P";
+      }
+
+      rule tag-teacher-members:
+      when {
+        $x isa person;
+        $y isa person;
+        (student: $x, teacher: $y) isa lesson;
+        (member: $x, member: $y, leader: $x) isa teams;
+      } then {
+        $y has tag "P";
+      }
+      """
+    Given for each session, graql insert
+      """
+      insert
+
+      $bob isa person, has name "bob";
+      $alice isa person, has name "alice";
+      $charlie isa person, has name "charlie";
+      $dennis isa person, has name "dennis";
+
+      (attendee: $bob, speaker: $alice) isa conference;
+      (member: $alice, member: $bob, host: $alice) isa party;
+
+      (attendee: $charlie, speaker: $dennis) isa conference;
+      (member: $charlie, member: $dennis, host: $charlie) isa party;
+      """
+    When materialised database is completed
       """
       match $x isa person, has name $n, has tag "P"; get;
       """
-  Then all answers are correct in reasoned database
-  Then answer size in reasoned database is: 2
-  Then materialised and reasoned databases are the same size
+    Then all answers are correct in reasoned database
+    Then answer size in reasoned database is: 2
+    Then materialised and reasoned databases are the same size
+
+  Scenario: when two distinct rules have alpha-equivalent bodies and heads, the reasoner still sees them as distinct.
+    Given for each session, graql define
+      """
+      define
+
+      person
+      plays husband,
+      plays wife;
+
+      marriage sub relation,
+      relates husband,
+      relates wife;
+
+      rule husbands-called-tracey:
+      when {
+          $x isa person;
+          $y isa person;
+          (husband: $x, wife: $y) isa marriage;
+      } then {
+          $x has name 'tracey';
+      };
+
+      rule: wives-called-tracey:
+      when {
+          $x isa person;
+          $y isa person;
+          (husband: $x, wife: $y) isa marriage;
+      } then {
+          $y has name 'tracey';
+      };
```

Comment:
Is the intention of this scenario to show that when the `when` blocks are both alpha equivalent, and the `then` blocks are both alpha equivalent, but when the `when` and `then considered together are clearly non-equivalent, that Reasoner still sees the two rules as distinct, and hence resolves both of them?

Reply:
Yes that is precisely the test. In 1.8.3 this doesn't pass, as it only checks if the `when` and `then` are alpha-equivalent.

Reply:
Extended the scenario description to make this clearer.